src/host/_stream.cpp

```diff
@@ -1053,36 +1052,25 @@ constexpr unsigned int LOCAL_BUFFER_SIZE = 100;
         const auto codepage = gci.OutputCP;
 
         // Convert our input parameters to Unicode
-        std::unique_ptr<wchar_t[]> wideCharBuffer{ nullptr };
-        static Utf8ToWideCharParser parser{ gci.OutputCP };

```

Comment:
**IGNORE, SEE BELOW** 
this one unfortunately cannot change just yet. `gci.OutputCP` _must_ be the user's console output codepage for now and the forseeable future.

This poses an interesting conundrum for u8u16: should we have "ASCII" versions that take a codepage? au16 and u16a? I'm not sure :smile:

Reply:
maybe it's actually part of the `u*state` itself...

Reply:
**OH**, I understand. You can almost entirely ignore this comment. I was confused because this was created outside the `CP_UTF8` block.
I also understand that writing in another codepage means we need to kill the u8 state -- so we can't move it inside this block.

Reply:
@DHowett-MSFT Your understanding is correct. The `reset()` method is called in the `else` branch where codepages other than UTF-8 are processed. I knew I would need the `reset` here, that's why I implemented it from the beginning.

> This poses an interesting conundrum for u8u16: should we have "ASCII" versions that take a codepage? au16 and u16a? I'm not sure 😄

Interesting indeed. Well, we already have function `ConvertToW` (and `ConvertToA`). But that's not applicable in `WriteConsoleAImpl()` due to the fact that we may receive DBCS-encoded text where caching of partials is required, too. @miniksa briefly mentioned that in https://github.com/Microsoft/terminal/issues/386#issuecomment-507823218
I only have a poor understanding of how DBCS has to be processed though. The manpage of `IsDBCSLeadByte` states that even if you validated a lead byte you may not rely on `MultiByteToWideChar` being able to process the substring correctly. So, unfortunately I don't know enough to revise the DBCS handling. And I don't know if there is any other function in the code base where DBCS has to be converted. Hence it might not worth the effort to bring it into a separate function.
However, I offer to do my best to get rid of `new` and `delete` in `WriteConsoleAImpl()` and instead use the wstring that we already have for the UTF-8 conversion.

Reply:
@german-one, if we theoretically made a `au16` and a `u16a` to supercede `ConvertToW` and `ConvertToA`, there would probably be some sort of `astate` variable required. That variable could be stored on the assorted input handles to ensure that the problem briefly discussed in #386 is rectified.

Off hand, I believe there are probably several points in the code base that could be unified behind such a convergence function in a similar way to converging the u8/u16 problem. But I haven't enumerated them recently, so I may be incorrect.

I believe that the only circumstance where we'd really pay attention to `IsDBCSLeadByte` and hold onto it until the next call is if a string ends in a lead byte (then it would take the lead byte off the end and cache it for the next call in the `astate` or equivalent.) If it's in the middle of a run, we wouldn't care and just pass it into `MultiByteToWideChar` probably with the replacement character flag on so it would remove invalid DBCS representations. The next write would have the stored DBCS lead prefixed to whatever comes next, even if it's not valid together, and we'd let MB2WC sort it out and replace it. 

The last provision to consider is that I believe the state would be discarded anytime the code page changed.

Reply:
@miniksa  `u16a` would be easy. We only need to take care of split surrogates, and that's what `u16state` already does.
As to `au16` - [`IsDBCSLeadByteEx`](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-isdbcsleadbyteex) isn't used in the current implementation of `CheckBisectStringA`. It would be my attempt to implement an `astate` though. The remarks found on the manpage still make me wonder if it would be bulletproof for DBCS.
But even if it was that simple, we would still only have conversions for SBCS, DBCS (the few mentioned on the manpage), and UTF-8. Guess what happens if we receive UTF-7 that was split inside of a base64 sequence 🤕